File: Packs/DemistoRESTAPI/Scripts/GetIncidentTasks/GetIncidentTasks.py
import demistomock as demisto  # noqa: F401
from CommonServerPython import *  # noqa: F401

from typing import Any


TASK_STATES = {
    'new': '',
    'inprogress': 'inprogress',
    'completed': 'Completed',
    'waiting': 'Waiting',
    'error': 'Error',
    'looperror': 'LoopError',
    'skipped': 'WillNotBeExecuted',
    'willnotbeexecuted': 'WillNotBeExecuted',
    'blocked': 'Blocked'
}


''' STANDALONE FUNCTION '''


def get_playbook_tasks(tasks: list) -> list:
    """Get the tasks of a playbook recursively

    Args:
        tasks (list): the tasks of the playbook

    Returns:
        list: the tasks fo the playbook including all sub-playbook tasks
    """
    ready_tasks = []
    for task in tasks:
        if task.get('type') == 'playbook' and task.get('subPlaybook'):
            sub_playbook_tasks = task.get("subPlaybook", {}).get("tasks", {}).values()
            ready_tasks.extend(get_playbook_tasks(list(sub_playbook_tasks)))
        ready_tasks.append(task)

    return ready_tasks


def get_states(states: list[str]) -> list[str]:
    """Get the internal states names for the given search states

    Args:
        states (list[str]): a list of states to search

    Returns:
        list[str]: a list of the internal names of the states given in the input or all possible states, if input was empty
    """
    if 'error' in states:
        states.append('loopError')

    ready_states = []
    for state in states:
        system_state = TASK_STATES.get(state.lower())
        if system_state and system_state not in ready_states:
            ready_states.append(system_state)

    if not ready_states:
        for state in TASK_STATES.values():
            if state not in ready_states:
                ready_states.append(state)

    return ready_states


def is_task_match(task: dict, name: str | None, tag: str | None, states: list[str]) -> bool:
    """Compares the given task with the given conditions to determine, if the Task matches

    Args:
        task (dict): The raw task
        name (str | None): the name to compare with the task or None, if not searching by name
        tag (str | None): the tag to compar with the task or None, if not searching by tag
        states (list[str]): a list of states that the the task may have

    Returns:
        bool: `True` if the task matches the given name, tag an/or state, otherwise `False`
    """
    task_task = task.get('task', {})
    name_match = name is None or task_task.get('name', '').lower() == name.lower()
    tag_match = tag is None or tag in task_task.get('tags', [])
    state_match = len(states) == 0 or task.get('state') in states

    return name_match and tag_match and state_match


''' COMMAND FUNCTION '''


def get_task_command(args: dict[str, Any]) -> CommandResults:
    name = args.get('name')
    tag = args.get('tag')
    states = get_states(argToList(args.get('states')))
    inc_id = args['inc_id']
    res = demisto.executeCommand('core-api-get', {'uri': f'/investigation/{inc_id}/workplan'})
    if not res or isError(res[0]):
        raise Exception(res)

    tasks: dict = dict_safe_get(res[0], ['Contents', 'response', 'invPlaybook', 'tasks'], {})
    if not tasks:
        return CommandResults(readable_output=f'Workplan for incident {inc_id} has no tasks.')

    allTasks = get_playbook_tasks(list(tasks.values()))
    res = []

    for task in allTasks:
        if is_task_match(task, name, tag, states):
            res.append({
                'id': task.get('id'),
                'name': task.get('task', {}).get('name', ''),
                'type': task.get('type'),
                'owner': task.get('assignee'),
                'state': task.get('state'),
                'scriptId': task.get('task', {}).get('scriptId'),
                'startDate': task.get('startDate'),
                'dueDate': task.get('dueDate'),
                'completedDate': task.get('completedDate'),
                'parentPlaybookID': task.get('parentPlaybookID'),
                'completedBy': task.get('completedBy')
            })

    return CommandResults(outputs_prefix='Tasks',
                          outputs_key_field='id',
                          outputs=res,
                          readable_output=tableToMarkdown(
                              name=f'Incident #{inc_id} Playbook Tasks',
                              t=res,
                              headers=['id', 'name', 'state', 'owner', 'scriptId'],
                              removeNull=True))


''' MAIN FUNCTION '''


def main():
    try:
        return_results(get_task_command(demisto.args()))
    except Exception as ex:
        return_error(f'Failed to execute Script Error: {str(ex)}')


''' ENTRY POINT '''


if __name__ in ('__main__', '__builtin__', 'builtins'):
    main()

File: Packs/DemistoRESTAPI/Scripts/GetIncidentTasks/GetIncidentTasks.js
TASK_STATES = {
    "new": "",
    "inprogress": "inprogress",
    "completed": "Completed",
    "waiting": "Waiting",
    "error": "Error",
    "looperror": "LoopError",
    "skipped": "WillNotBeExecuted",
    "willnotbeexecuted": "WillNotBeExecuted",
    "blocked": "Blocked",
    undefined: undefined
}


function getPlaybookTasks(tasks){
    var readyTasks = [];

    for (var i = 0; i < tasks.length; i++) {
        var task = tasks[i];

        if (task.type === "playbook" && task.subPlaybook) {
            var subPlaybookTasks = [];

            if (task.subPlaybook.tasks) {
                for (var key in task.subPlaybook.tasks) {
                    if (task.subPlaybook.tasks.hasOwnProperty(key)) {
                        subPlaybookTasks.push(task.subPlaybook.tasks[key]);
                    }
                }
            }

            var nestedTasks = getPlaybookTasks(subPlaybookTasks);
            readyTasks = readyTasks.concat(nestedTasks);
        }

        readyTasks.push(task);
    }

    return readyTasks;
}


function isTaskMatch(task, name, tag, states) {
    var taskTask = task.task || {};

    var nameMatch = name === null || name === undefined ||
        (taskTask.name && taskTask.name.toLowerCase() === name.toLowerCase());

    var tagMatch = tag === null || tag === undefined ||
        (taskTask.tags && taskTask.tags.indexOf(tag) !== -1);

    var stateMatch = !states || states.length === 0 ||
        (states.indexOf(task.state) !== -1);

    return nameMatch && tagMatch && stateMatch;
}


function getStates(states) {
    // If "error" is in the input list, add "loopError"
    if (states.indexOf("error") !== -1) {
        states.push("loopError");
    }

    var readyStates = [];

    for (var i = 0; i < states.length; i++) {
        var state = states[i];
        var systemState = TASK_STATES[state.toLowerCase()];

        if (systemState && readyStates.indexOf(systemState) === -1) {
            readyStates.push(systemState);
        }
    }

    // If no matches were found, return all values from TASK_STATES
    if (readyStates.length === 0) {
        for (var key in TASK_STATES) {
            if (TASK_STATES.hasOwnProperty(key)) {
                var value = TASK_STATES[key];
                if (readyStates.indexOf(value) === -1) {
                    readyStates.push(value);
                }
            }
        }
    }

    return readyStates;
}

var name = args.name
var tag = args.tag
var inc_id = args.inc_id
var states = getStates(argToList(args.states))

var res = executeCommand('core-api-get', {
    uri: `/investigation/${inc_id}/workplan`});


if (!isValidRes(res))
{
    if(res[0].Contents){
          return {
              ContentsFormat: formats.markdown,
              Type: entryTypes.error,
              Contents: res[0].Contents
          };
    }
}

var tasks = undefined;
if (
    res &&
    res[0] &&
    res[0].Contents &&
    res[0].Contents.response &&
    res[0].Contents.response.invPlaybook &&
    res[0].Contents.response.invPlaybook.tasks
)
{
    tasks = res[0].Contents.response.invPlaybook.tasks;


    if (tasks == undefined){
        return{
            Type: entryTypes.note,
            ReadableContentsFormat: formats.text,
            Contents: `Workplan for incident ${inc_id} has no tasks.`,
        };
    }

    allTasks = getPlaybookTasks(Object.values(tasks));


    var res = [];

    for (var i = 0; i < allTasks.length; i++) {
        var task = allTasks[i];

        if (isTaskMatch(task, name, tag, states)) {
            res.push({
                id: task.id,
                name: (task.task && task.task.name) || "",
                type: task.type,
                owner: task.assignee,
                state: task.state,
                scriptId: task.task && task.task.scriptId,
                startDate: task.startDate,
                dueDate: task.dueDate,
                completedDate: task.completedDate,
                parentPlaybookID: task.parentPlaybookID,
                completedBy: task.completedBy
            });
        }
    }


    var ec = {'Tasks(val.id && val.id == obj.id)': res};


    return {
        Type: entryTypes.note,
        ReadableContentsFormat: formats.markdown,
        Contents: res,
        ContentsFormat: formats.json,
        HumanReadable: tableToMarkdown(`Incident #${inc_id} Playbook Tasks`, res, ["id", "name", "state", "owner", "scriptId"],removeNull=true),
        EntryContext: ec
    };

}




if (tasks == undefined){
    return{
        Type: entryTypes.note,
        ReadableContentsFormat: formats.text,
        Contents: `Workplan for incident ${inc_id} has no tasks.`,
    };
}
